adapters/postgres.go

```diff
@@ -0,0 +1,56 @@
+package adapters
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+
+	"github.com/pkg/errors"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+type Database struct {
+	db *gorm.DB
+	in *Config
+}
+
+func NewDatabase(in *Config) *Database {
+	return &Database{
+		in: in,
+	}
+}
+
+func (d *Database) AutoMigration(arg ...interface{}) error {
+	err := d.db.AutoMigrate(arg...)
+
+	if err != nil {
+		return errors.Wrap(err, "Couldnt migrate database structs")
+	}
+
+	return nil
+}
+
+func (d *Database) StartPostgres() {
+
+	dns := fmt.Sprintf(
+		"host=%s user=%s password=%s dbname=%s port=%s sslmode=disable",
+		d.in.Host,
+		d.in.User,
+		d.in.Password,
+		d.in.DbName,
+		d.in.Port,
+	)
+
+	database, err := gorm.Open(postgres.Open(dns), &gorm.Config{})
+	if err != nil {
+		fmt.Println("Could not connect to the Postgres Database")
+		log.Fatal("error: ", err)
+	}
+
+	d.db = database
+}
+
+func (d *Database) GetDatabase() (*sql.DB, error) {
+	return d.db.DB()
+}
```

Comment:
O que acham de fazer mais ou menos assim?

```suggestion
func NewDatabase(in *Config) *gorm.DB {
	dsn := fmt.Sprintf(
		"host=%s user=%s password=%s dbname=%s port=%s sslmode=disable",
		in.Host,
		in.User,
		in.Password,
		in.DbName,
		in.Port,
	)

	database, err := gorm.Open(postgres.Open(dsn), &gorm.Config{})
	if err != nil {
		fmt.Println("Could not connect to the Postgres Database")
		log.Fatal("error: ", err)
	}

	return database
}
```

* O principal aqui é que antes tinha uma dependência implicita de ordem, por exemplo, se vc chamasse o `AutoMigrate` antes do `StartPostgres`...ia estourar! Agora isso é simplesmente impossível de acontecer.
* `NewDatabase` te retona o db do gorm, que é o que vcs precisam pra passar para os repos, isso tira a necessidade de ter o `GetDatabse`;
* O `AutoMigrate`, honestamente, eu acho mais interessante só chamar na main do `migrate` direto;

No geral eu acho esse pacote `adapters` desnecessário, mas se forem manter ele, eu diria que essa forma fica um pouco melhor.

O que acham?

Reply:
Bom demais, dessa maneira enxugou muito o código! :nerd_face: